### PR TITLE
PTFE-672: api authentication

### PR DESCRIPTION
**Pull request by Abubakarr99**

### CODE REVIEW

File: runner_manager/main.py [hunk 1]
```diff
@@ -1,13 +1,39 @@
...
+def get_api_key(
+    api_key_query: str = Security(api_key_query),
+    api_key_header: str = Security(api_key_header),
+) -> str:
+    settings = get_settings()
+    if not settings.api_key:
+        raise HTTPException(
+            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
+            detail="API Key not configured in settings",
+        )
```
Comment (tcarmet) [inline runner_manager/main.py:24]: If we haven't configured the api key then we can proceed with the request without making any kind of authentication.

```suggestion
    if not settings.api_key:
        return settings.api_key
```

File: runner_manager/main.py [hunk 2]
```diff
@@ -1,13 +1,39 @@
...
+        raise HTTPException(
+            status_code=status.HTTP_401_UNAUTHORIZED,
+            detail="API Key required for this endpoint",
+        )
+    if api_key != settings.api_key.get_secret_value():
+        raise HTTPException(
+            status_code=status.HTTP_401_UNAUTHORIZED,
+            detail="Invalid API Key",
+        )
+    return api_key
```
Comment (tcarmet) [inline runner_manager/main.py:36]:
```suggestion
    if api_key_query in [settings.api_key]:
        return api_key_query
    if api_key_header in [settings.api_key]:
        return api_key_header
    raise HTTPException(
        status_code=status.HTTP_401_UNAUTHORIZED,
        detail="Invalid API Key",
    )
```

**Review by tcarmet — COMMENTED** [review]
This is how I see it according to the example.

File: tests/api/tests_auth.py [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    response = client.get("/private")
+    # for now we are not using api key
+    assert response.status_code == 200
+
+
+def test_private_endpoint_with_valid_api_key(client):
+    os.environ["API_KEY"] = "secret"
+    headers = {"x-api-key": "secret"}
+    response = client.get("/private", headers=headers)
+    assert response.status_code == 200
```
Comment (tcarmet) [inline tests/api/tests_auth.py:20]:
```suggestion
@lru_cache()
def settings_api_key():
    return Settings(api_key="secret")


def test_private_endpoint_with_valid_api_key(fastapp, client):
    fastapp.dependency_overrides[get_settings] = settings_api_key

    headers = {"x-api-key": "secret"}
    with pytest.raises(HttpError403): # to be fixed wit accurate exception name
        client.get("/private")
    response = client.get("/private", headers=headers)
    assert response.status_code == 200
```

If it doesn't work, review this file for inspiration: https://github.com/scality/gh-actions-exporter/blob/main/tests/api/metrics/conftest.py

File: runner_manager/main.py [hunk 4]
```diff
@@ -1,14 +1,35 @@
...
 app = FastAPI()
+api_key_query = APIKeyQuery(name="api-key", auto_error=False)
+api_key_header = APIKeyHeader(name="x-api-key", auto_error=False)
+
+
+def get_api_key(
+    api_key_query: str = Security(api_key_query),
+    api_key_header: str = Security(api_key_header),
+) -> str:
+    settings = get_settings()
```
Comment (tcarmet) [inline runner_manager/main.py:21]:
```suggestion
def get_api_key(
    api_key_query: str = Security(api_key_query),
    api_key_header: str = Security(api_key_header),
    settings: Settings = Depends(get_settings),
) -> str:
```

File: runner_manager/main.py [hunk 5]
```diff
@@ -1,14 +1,35 @@
 import logging
 
-from fastapi import FastAPI, Response
+from fastapi import FastAPI, HTTPException, Response, Security, status
```
Comment (tcarmet) [inline runner_manager/main.py:3]:
```suggestion
from fastapi import FastAPI, HTTPException, Response, Security, status, Depends
```

File: runner_manager/main.py [hunk 6]
```diff
@@ -1,14 +1,35 @@
 import logging
 
-from fastapi import FastAPI, Response
+from fastapi import FastAPI, HTTPException, Response, Security, status, Depends
+from fastapi.security import APIKeyHeader, APIKeyQuery
 
-from runner_manager.dependencies import get_queue
+from runner_manager.dependencies import get_queue, get_settings
```
Comment (tcarmet) [inline runner_manager/main.py:6]:
```suggestion
from runner_manager.models.settings import Settings
from runner_manager.dependencies import get_queue, get_settings
```

File: .github/workflows/tests.yaml [hunk 7]
```diff
@@ -27,7 +27,7 @@ jobs:
     env:
       REDIS_OM_URL: redis://localhost:6379/0
       GITHUB_BASE_URL: http://localhost:4010
-
+      API_KEY: ${{ secrets.API_KEY }}
```
Comment (tcarmet) [inline .github/workflows/tests.yaml:30]: shouldn't be set
```suggestion
```

**Review by tcarmet — CHANGES_REQUESTED** [review]

File: runner_manager/main.py [hunk 8]
```diff
@@ -1,14 +1,36 @@
...
 from runner_manager.jobs.startup import startup
+from runner_manager.models.settings import Settings
 from runner_manager.routers import webhook
 
 log = logging.getLogger(__name__)
 
 app = FastAPI()
+api_key_query = APIKeyQuery(name="api-key", auto_error=False)
+api_key_header = APIKeyHeader(name="x-api-key", auto_error=False)
+
```
[the remark this answers is not in the document]
Comment (Abubakarr99) [reply runner_manager/main.py:18]: ok

File: runner_manager/main.py [hunk 9]
```diff
@@ -21,6 +25,13 @@ def startup_event():
...
+@app.get("/public")
+def public():
+    """A public endpoint that does not require any authentication."""
+    return "Public Endpoint"
+
+
+@app.get("/private")
+def private(api_key: str = Security(get_api_key)):
+    """A private endpoint that requires a valid API key to be provided."""
+    return f"Private Endpoint. API Key: {api_key}"
```
Comment (tcarmet) [inline runner_manager/main.py:37]: We have no use for those for now, but they are usefull for testing, would you mind trying to set them as a router inside `test_auth.py` (check `routers/_health.py` for inspiration)

Then in your test include `fastapp` `def test_my_test(fastapp, ...):`

Then proceed in adding those routes for the tests only: `fastapp.include_router(my_auth_test_router)` then execute the tests.

**Review by tcarmet — APPROVED** [review]
It LGTM, left one task that is harder to do speaking about routes, try your best on it, if it doesn't work out of the box leave it as is and we'll re-iterate on it as needed.

**Pull request merged**